File: group-anagrams/group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> group = new HashMap<>();
        
        for(String str : strs) {
            List<String> list = group.computeIfAbsent(encode(str), k -> new ArrayList<>());
            list.add(str);
        }
        
        List<List<String>> result = new ArrayList<>();
        
        for(Map.Entry<String, List<String>> kv : group.entrySet()) {
            result.add(kv.getValue());
        }
        
        return result;
    }
    
    String encode(String str) {
        int[] frequency = new int[26];
        
        for(char c : str.toCharArray()) {
            frequency[(int)c - 'a']++;
        }
        
        StringBuilder sb = new StringBuilder();
        
        for(int i = 0; i < 26; i++) {
            sb.append(frequency[i] + " ");
        }
        
        return sb.toString();
    }
}

File: group-anagrams/group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        m = {}
        
        for s in strs:
            enc = self.encode(s)
            m[enc] = m.get(enc, []) + [s]
                
        return list(m.values())    
        
    def encode(self, s: str) -> str:
        bkt = [0] * 26
        b = ord('a')
        for c in s: bkt[ord(c) - b] += 1
        return str(bkt)
        